### PR TITLE
build: reduce iamge size by removing unused packages

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: https://github.com/longhorn/longhorn/issues/5911

### CODE REVIEW

**Comment by ChanYiLin** [comment]
> This pull request is now in conflicts. Could you fix it @ChanYiLin? 🙏

conflicts resolved

**Comment by ChanYiLin** [comment]
e2e tests: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5257/

File: package/Dockerfile [hunk 1]
```diff
@@ -77,11 +81,9 @@ RUN zypper -n addrepo --refresh https://download.opensuse.org/repositories/syste
     zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:backports/SLE_15/devel:languages:python:backports.repo && \
     zypper --gpg-auto-import-keys ref
 
-RUN zypper -n install kmod curl fuse wget \
-    nfs-client nfs4-acl-tools cifs-utils \
-    libibverbs librdmacm1 rdma-core-devel perl-Config-General libaio-devel sg3_utils \
-    iputils telnet iproute2 qemu-tools \
-    e2fsprogs e2fsprogs-devel xfsprogs xfsprogs-devel util-linux-systemd python3
+RUN zypper -n install nfs-client nfs4-acl-tools cifs-utils libaio1 sg3_utils \
+    iputils iproute2 qemu-tools e2fsprogs util-linux-systemd python3 && \
```
Comment (derekbit) [inline package/Dockerfile:85]: Is python3 necessary?

Comment (ChanYiLin) [reply package/Dockerfile]: Removed, thanks!

**Comment by ChanYiLin** [comment]
> In general, LGTM.
> 
> Just have the same question as Derek, python3 runtime should not required.

Hi @innobead , removed, thanks!

**Review by derekbit — APPROVED** [review]

File: package/Dockerfile [hunk 2]
```diff
@@ -77,11 +81,9 @@
     zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:backports/SLE_15/devel:languages:python:backports.repo && \
     zypper --gpg-auto-import-keys ref
 
-RUN zypper -n install kmod curl fuse wget \
-    nfs-client nfs4-acl-tools cifs-utils \
-    libibverbs librdmacm1 rdma-core-devel perl-Config-General libaio-devel sg3_utils \
-    iputils telnet iproute2 qemu-tools \
-    e2fsprogs e2fsprogs-devel xfsprogs xfsprogs-devel util-linux-systemd python3
+RUN zypper -n install nfs-client nfs4-acl-tools cifs-utils libaio1 sg3_utils \
```
Comment (innobead) [inline package/Dockerfile:84]: `libaio1` will be used at runtime?

**Review by innobead — CHANGES_REQUESTED** [review]
Do we need the following repos in the image?

```bash
    zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:libraries:c_c++/15.4/devel:libraries:c_c++.repo && \
    zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:Factory/15.4/devel:languages:python:Factory.repo && \
    zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:backports/SLE_15/devel:languages:python:backports.repo && \
```

**Comment by ChanYiLin** [comment]
> Do we need the following repos in the image?
> 
> ```shell
>     zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:libraries:c_c++/15.4/devel:libraries:c_c++.repo && \
>     zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:Factory/15.4/devel:languages:python:Factory.repo && \
>     zypper -n addrepo --refresh https://download.opensuse.org/repositories/devel:languages:python:backports/SLE_15/devel:languages:python:backports.repo && \
> ```


These are added when introducing spdk, don't know if these are needed in the runtime, may meed to ask @derekbit 
https://github.com/longhorn/longhorn-instance-manager/commit/0225d0454aeba17ea2edfa2b34d900e0dbf34e3a

**Comment by ChanYiLin** [comment]
oh wait those are just adding repo
I think we can remove them since we are not installing these package in runtime

File: package/Dockerfile (hunk 2 above)
Comment (ChanYiLin) [reply package/Dockerfile]: removed

**Review by innobead — APPROVED** [review]

**Comment by derekbit** [comment]
Don’t merge. The spdk that cannot be started after applying the PR.

取得 iOS 版 Outlook<https://aka.ms/o0ukef>
________________________________
寄件者: David Ko ***@***.***>
寄件日期: Friday, November 17, 2023 4:44:15 PM
收件者: longhorn/longhorn-instance-manager ***@***.***>
副本: Derek Su ***@***.***>; Mention ***@***.***>
主旨: Re: [longhorn/longhorn-instance-manager] build: reduce iamge size by removing unused packages (PR #311)


@innobead approved this pull request.

**Comment by derekbit** [comment]
Ref 

http://github.com/derekbit/longhorn-instance-manager/commit/55cb84a77c080a66d01d146aeab4b565e0b7098d

cc @ChanYiLin

**Comment by ChanYiLin** [comment]
Hi @derekbit @innobead 
I have tested the PR
- amd64 coretest
- arm64 coretest
- install longhorn and try spdk volume function (attach to pods)

Do we still have any concerns I missed?

**Review by innobead — APPROVED** [review]
LGTM.

However, we need to review if we still need the following dependenceies at the runtime image, because it seems installing unnecessary dependdencies for building SPDK artificats, even though some runtime libs we do need. 

```bash
# Install SPDK dependencies
COPY --from=builder /usr/src/spdk/scripts /spdk/
RUN if [ "${ARCH}" == "amd64" ]; then \
        sh /spdk/pkgdep.sh; \
        rm -rf /tmp/spdk; \
    fi
```

**Comment by derekbit** [comment]
@ChanYiLin 
Can you replace `rm -rf /tmp/spdk; \` with `rm -rf /spdk; \`? Thank you.

**Comment by derekbit** [comment]
> LGTM.
> 
> However, we need to review if we still need the following dependencies at the runtime image, because it seems to install unnecessary dependencies for building SPDK artifacts, even though we need some runtime libs.
> 
> ```shell
> # Install SPDK dependencies
> COPY --from=builder /usr/src/spdk/scripts /spdk/
> RUN if [ "${ARCH}" == "amd64" ]; then \
>         sh /spdk/pkgdep.sh; \
>         rm -rf /tmp/spdk; \
>     fi
> ```

After a quick check, pkgdeps.sh installs the minimal dependencies because we don't enable unnecessary options.
However, the four tools are not required.
https://github.com/spdk/spdk/blob/master/scripts/pkgdep/sles.sh#L14

**Comment by ChanYiLin** [comment]
> @ChanYiLin Can you replace `rm -rf /tmp/spdk; \` with `rm -rf /spdk; \`? Thank you.

Hi @derekbit fixed, thanks!

**Review by derekbit — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Review by derekbit — APPROVED** [review]

**Comment by innobead** [comment]
Not sure why can't rebase and then merge the PR. IIRC, there were several times when merging your PRs. @ChanYiLin , probably need to check how to resolve this next time.

**Comment by derekbit** [comment]
> Not sure why can't rebase and then merge the PR. IIRC, there were several times when merging your PRs. @ChanYiLin , probably need to check how to resolve this next time.

https://github.com/orgs/community/discussions/27521

**Comment by ChanYiLin** [comment]
@mergify backport v1.5.x

**Pull request merged**